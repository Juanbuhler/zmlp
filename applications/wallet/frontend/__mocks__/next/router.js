/**
 * useRouter
 */

let mockUseRouter = {
  query: {
    action: '',
  },
}

export const __setUseRouter = data => {
  mockUseRouter = data
}

export const useRouter = () => mockUseRouter

/**
 * push
 */

let mockPushFunction = () => {}

export const __setMockPushFunction = fn => {
  mockPushFunction = fn
}

/**
 * on
 */

let mockOnFunction = () => {}

export const __setMockOnFunction = fn => {
  mockOnFunction = fn
}

/**
 * off
 */

let mockOffFunction = () => {}

export const __setMockOffFunction = fn => {
  mockOffFunction = fn
}

/**
 * Router
 */

const Router = {
  push: (...args) => {
    mockPushFunction(...args)
  },
  events: {
    on: (...args) => {
      mockOnFunction(...args)
    },
    off: (...args) => {
      mockOffFunction(...args)
    },
  },
}

export default Router
