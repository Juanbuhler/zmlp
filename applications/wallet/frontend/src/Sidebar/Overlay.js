import React from 'react'
import PropTypes from 'prop-types'

import { zIndex, colors } from '../Styles'

const SidebarOverlay = ({ isSidebarOpen, setSidebarOpen }) => (
  <div
    role="button"
    aria-label="Close Sidebar Menu"
    tabIndex="-1"
    css={{
      position: 'fixed',
      top: 0,
      left: 0,
      bottom: 0,
      zIndex: zIndex.layout.overlay,
      backgroundColor: colors.structure.black,
      opacity: isSidebarOpen ? 0.5 : 0,
      width: isSidebarOpen ? '100%' : 0,
      transition: isSidebarOpen
        ? 'opacity ease-in-out .3s, width ease-in 0s 0s'
        : 'opacity ease-in-out .3s, width ease-in 0s .3s',
    }}
    onClick={() => setSidebarOpen(false)}
    onKeyDown={() => setSidebarOpen(false)}
  />
)

SidebarOverlay.propTypes = {
  isSidebarOpen: PropTypes.bool.isRequired,
  setSidebarOpen: PropTypes.func.isRequired,
}

export default SidebarOverlay
