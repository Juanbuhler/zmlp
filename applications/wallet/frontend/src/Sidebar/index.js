import { forwardRef, useEffect } from 'react'
import PropTypes from 'prop-types'
import Router from 'next/router'

import { colors, spacing, zIndex, constants } from '../Styles'

import SidebarLink from './Link'
import SidebarOverlay from './Overlay'

import ProjectDashboardSvg from '../Icons/projectDashboard.svg'
import DataSourcesSvg from '../Icons/datasources.svg'
import JobQueueSvg from '../Icons/jobQueue.svg'
import VisualizerSvg from '../Icons/visualizer.svg'
import KeySvg from '../Icons/key.svg'
import UsersSvg from '../Icons/users.svg'
import AccountDashboardSvg from '../Icons/accountDashboard.svg'
import GearSvg from '../Icons/gear.svg'

const WIDTH = 240
const ICON_WIDTH = 20

const Sidebar = forwardRef(
  ({ projectId, isSidebarOpen, setSidebarOpen }, ref) => {
    useEffect(() => {
      const handleRouteChange = () => {
        setSidebarOpen(false)
      }

      Router.events.on('routeChangeStart', handleRouteChange)

      return () => {
        Router.events.off('routeChangeStart', handleRouteChange)
      }
    }, [setSidebarOpen])

    return (
      <div>
        <nav
          ref={ref}
          css={{
            width: WIDTH,
            position: 'fixed',
            height: `calc(100% - ${constants.navbar.height}px)`,
            overflowY: 'auto',
            zIndex: zIndex.layout.drawer,
            backgroundColor: colors.structure.iron,
            transition: 'left ease-in-out .3s, visibility ease-in-out .3s',
            overscrollBehavior: 'contain',
            left: isSidebarOpen ? 0 : -WIDTH,
            top: constants.navbar.height,
            display: 'flex',
            flexDirection: 'column',
            justifyContent: 'space-between',
          }}
        >
          <ul
            css={{
              listStyleType: 'none',
              padding: 0,
              margin: 0,
              paddingTop: spacing.moderate,
              paddingBottom: spacing.moderate,
            }}
          >
            <SidebarLink projectId={projectId} href="/[projectId]">
              <ProjectDashboardSvg width={ICON_WIDTH} aria-hidden />
              Project Dashboard
            </SidebarLink>

            <SidebarLink projectId={projectId} href="/[projectId]/data-sources">
              <DataSourcesSvg width={ICON_WIDTH} aria-hidden />
              Data Sources
            </SidebarLink>

            <SidebarLink projectId={projectId} href="/[projectId]/jobs">
              <JobQueueSvg width={ICON_WIDTH} aria-hidden />
              Job Queue
            </SidebarLink>

            <SidebarLink projectId={projectId} href="/[projectId]/visualizer">
              <VisualizerSvg width={ICON_WIDTH} aria-hidden />
              Visualizer
            </SidebarLink>

            <SidebarLink projectId={projectId} href="/[projectId]/api-keys">
              <KeySvg width={ICON_WIDTH} aria-hidden />
              API Keys
            </SidebarLink>

            <SidebarLink projectId={projectId} href="/[projectId]/users">
              <UsersSvg width={ICON_WIDTH} aria-hidden />
              User Admin
            </SidebarLink>
          </ul>
          <ul
            css={{
              listStyleType: 'none',
              padding: 0,
              margin: 0,
              paddingTop: spacing.moderate,
              paddingBottom: spacing.moderate,
              backgroundColor: colors.structure.smoke,
            }}
          >
            <SidebarLink projectId={projectId} href="/">
              <AccountDashboardSvg width={ICON_WIDTH} aria-hidden />
              Account Dashboard
            </SidebarLink>

            <SidebarLink projectId={projectId} href="/account">
              <GearSvg width={ICON_WIDTH} aria-hidden />
              Account
            </SidebarLink>
          </ul>
        </nav>
        <SidebarOverlay
          isSidebarOpen={isSidebarOpen}
          setSidebarOpen={setSidebarOpen}
        />
      </div>
    )
  },
)

Sidebar.propTypes = {
  projectId: PropTypes.string.isRequired,
  isSidebarOpen: PropTypes.bool.isRequired,
  setSidebarOpen: PropTypes.func.isRequired,
}

export default Sidebar
