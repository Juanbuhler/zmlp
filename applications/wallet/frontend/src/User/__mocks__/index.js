import { createContext, createElement } from 'react'

import user from './user'

export const noop = () => {}

export const UserContext = createContext({
  user,
  googleAuth: {},
  setGoogleAuth: noop,
})

const User = ({ children, ...rest }) => createElement('User', rest, children)

export default User
