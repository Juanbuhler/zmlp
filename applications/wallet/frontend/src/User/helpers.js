export const noop = () => {}

export const meFetcher =
  typeof window === 'undefined'
    ? noop
    : async (url) => {
        try {
          const response = await fetch(url)
          if (response.status >= 400) throw response
          return response.json()
        } catch (error) {
          return {}
        }
      }
