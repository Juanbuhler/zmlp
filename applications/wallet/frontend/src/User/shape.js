import PropTypes from 'prop-types'

const userShape = {
  id: PropTypes.number.isRequired,
  username: PropTypes.string.isRequired,
  email: PropTypes.string.isRequired,
  firstName: PropTypes.string.isRequired,
  lastName: PropTypes.string.isRequired,
  organizations: PropTypes.arrayOf(PropTypes.string).isRequired,
}

export default userShape
