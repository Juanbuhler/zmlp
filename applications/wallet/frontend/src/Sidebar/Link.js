import PropTypes from 'prop-types'
import Link from 'next/link'
import { useRouter } from 'next/router'

import { colors, constants, spacing, typography } from '../Styles'

const SidebarLink = ({ projectId, href, children }) => {
  const { pathname } = useRouter()

  const isCurrentPage = pathname === href

  return (
    <li css={{ borderBottom: constants.borders.regular.transparent }}>
      <Link href={href} as={href.replace('[projectId]', projectId)} passHref>
        <a
          css={{
            display: 'flex',
            alignItems: 'center',
            padding: spacing.moderate,
            fontSize: typography.size.regular,
            lineHeight: typography.height.regular,
            backgroundColor: isCurrentPage ? colors.structure.mattGrey : 'none',
            color: isCurrentPage ? colors.key.two : colors.structure.zinc,
            svg: { marginRight: spacing.moderate },
            ':hover': {
              textDecoration: 'none',
              color: colors.structure.white,
              backgroundColor: colors.structure.mattGrey,
            },
          }}
        >
          {children}
        </a>
      </Link>
    </li>
  )
}

SidebarLink.propTypes = {
  projectId: PropTypes.string.isRequired,
  href: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
}

export default SidebarLink
