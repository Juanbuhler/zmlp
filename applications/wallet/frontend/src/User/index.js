import { createContext, useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import * as Sentry from '@sentry/browser'

import { getUser, setUser } from './helpers'

export const noop = () => {}

export const UserContext = createContext({
  user: {},
  setUser: noop,
  googleAuth: {},
  setGoogleAuth: noop,
})

const User = ({ initialUser, children }) => {
  const [user, setStateUser] = useState(initialUser)
  const [hasLocalStorageLoaded, setHasLocalStorageLoaded] = useState(false)
  const [googleAuth, setGoogleAuth] = useState({ signIn: noop, signOut: noop })

  useEffect(() => {
    if (initialUser.id || hasLocalStorageLoaded) return

    const storedUser = getUser()

    setStateUser(storedUser)

    setHasLocalStorageLoaded(true)

    /* istanbul ignore next */
    Sentry.configureScope(scope => {
      scope.setUser(storedUser)
    })
  }, [initialUser, hasLocalStorageLoaded, user, setStateUser])

  if (!initialUser.id && !hasLocalStorageLoaded) return null

  return (
    <UserContext.Provider
      value={{
        user,
        setUser: setUser({ setStateUser, user }),
        googleAuth,
        setGoogleAuth,
      }}>
      {children}
    </UserContext.Provider>
  )
}

User.propTypes = {
  initialUser: PropTypes.shape({
    id: PropTypes.number,
    email: PropTypes.string,
  }).isRequired,
  children: PropTypes.node.isRequired,
}

export default User
