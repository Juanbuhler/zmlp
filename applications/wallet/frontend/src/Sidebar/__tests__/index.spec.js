import TestRenderer, { act } from 'react-test-renderer'

import mockUser from '../../User/__mocks__/user'

import User from '../../User'

import Sidebar from '..'

const noop = () => () => {}

const PROJECT_ID = '76917058-b147-4556-987a-0a0f11e46d9b'

describe('<Sidebar />', () => {
  it('should render properly closed', () => {
    require('next/router').__setUseRouter({
      pathname: '/[projectId]/jobs',
    })

    const component = TestRenderer.create(
      <User initialUser={mockUser}>
        <Sidebar
          projectId={PROJECT_ID}
          isSidebarOpen={false}
          setSidebarOpen={noop}
        />
      </User>,
    )

    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should render properly opened', () => {
    require('next/router').__setUseRouter({
      pathname: '/[projectId]/jobs',
    })

    const component = TestRenderer.create(
      <User initialUser={mockUser}>
        <Sidebar projectId={PROJECT_ID} isSidebarOpen setSidebarOpen={noop} />
      </User>,
    )

    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should close when clicking on the overlay', () => {
    const mockFn = jest.fn()

    const component = TestRenderer.create(
      <Sidebar projectId={PROJECT_ID} isSidebarOpen setSidebarOpen={mockFn} />,
    )

    act(() => {
      component.root
        .findByProps({ 'aria-label': 'Close Sidebar Menu' })
        .props.onClick({ preventDefault: noop })
    })

    expect(mockFn).toHaveBeenCalledWith(false)
  })

  it('should close when key pressing on the overlay', () => {
    const mockFn = jest.fn()

    const component = TestRenderer.create(
      <Sidebar projectId={PROJECT_ID} isSidebarOpen setSidebarOpen={mockFn} />,
    )

    act(() => {
      component.root
        .findByProps({ 'aria-label': 'Close Sidebar Menu' })
        .props.onKeyDown({ preventDefault: noop })
    })

    expect(mockFn).toHaveBeenCalledWith(false)
  })

  it('should close the sidebar on route change', () => {
    const mockOnFunction = jest.fn((_, callback) => {
      // invoke callbacks immediately instead of queuing them
      callback()
    })
    const mockSetSidebarOpen = jest.fn()
    const mockOffFunction = jest.fn()

    require('next/router').__setMockOnFunction(mockOnFunction)
    require('next/router').__setMockOffFunction(mockOffFunction)

    const component = TestRenderer.create(
      <Sidebar
        projectId={PROJECT_ID}
        isSidebarOpen
        setSidebarOpen={mockSetSidebarOpen}
      />,
    )

    // useEffect
    act(() => {})

    expect(mockOnFunction).toHaveBeenCalled()
    expect(mockSetSidebarOpen).toHaveBeenCalled()

    act(() => {
      component.unmount()
    })

    expect(mockOffFunction).toHaveBeenCalled()
  })
})
