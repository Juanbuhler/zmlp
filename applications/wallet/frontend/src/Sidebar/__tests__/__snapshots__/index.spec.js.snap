// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar /> should render properly closed 1`] = `
.emotion-9 {
  width: 240px;
  position: fixed;
  height: calc(100% - 44px);
  overflow-y: auto;
  z-index: 90;
  background-color: #4a4a4a;
  padding-bottom: 32px;
  -webkit-transition: left ease-in-out .3s,visibility ease-in-out .3s;
  transition: left ease-in-out .3s,visibility ease-in-out .3s;
  overscroll-behavior: contain;
  left: -240px;
  top: 44px;
  padding-top: 12px;
}

.emotion-8 {
  list-style-type: none;
  padding: 0;
  margin: 0;
}

.emotion-1 {
  border-bottom: 1px solid transparent;
}

.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 12px;
  font-size: 15px;
  line-height: 18px;
  background-color: #3b3b3b;
  color: #73b61c;
}

.emotion-0 svg {
  margin-right: 12px;
}

.emotion-0:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #73b61c;
  background-color: #3b3b3b;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 12px;
  font-size: 15px;
  line-height: 18px;
  background-color: none;
  color: #808080;
}

.emotion-2 svg {
  margin-right: 12px;
}

.emotion-2:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #e0e0e0;
  background-color: #3b3b3b;
}

.emotion-10 {
  position: fixed;
  top: 0;
  left: 0;
  bottom: 0;
  z-index: 80;
  background-color: #000000;
  opacity: 0;
  width: 0;
  -webkit-transition: opacity ease-in-out .3s,width ease-in 0s .3s;
  transition: opacity ease-in-out .3s,width ease-in 0s .3s;
}

<div>
  <nav
    className="emotion-9"
  >
    <ul
      className="emotion-8"
    >
      <li
        className="emotion-1"
      >
        <a
          className="emotion-0"
          href="/76917058-b147-4556-987a-0a0f11e46d9b/jobs"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <svg
            aria-hidden={true}
            data-file-name="queue.svg"
            width={20}
          />
          Job Queue
        </a>
      </li>
      <li
        className="emotion-1"
      >
        <a
          className="emotion-2"
          href="/76917058-b147-4556-987a-0a0f11e46d9b/api-keys"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <svg
            aria-hidden={true}
            data-file-name="key.svg"
            width={20}
          />
          API Keys
        </a>
      </li>
      <li
        className="emotion-1"
      >
        <a
          className="emotion-2"
          href="/76917058-b147-4556-987a-0a0f11e46d9b/users"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <svg
            aria-hidden={true}
            data-file-name="users.svg"
            width={20}
          />
          User Admin
        </a>
      </li>
      <li
        className="emotion-1"
      >
        <a
          className="emotion-2"
          href="/76917058-b147-4556-987a-0a0f11e46d9b/account"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <svg
            aria-hidden={true}
            data-file-name="gear.svg"
            width={20}
          />
          User Admin
        </a>
      </li>
    </ul>
  </nav>
  <div
    aria-label="Close Sidebar Menu"
    className="emotion-10"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="-1"
  />
</div>
`;

exports[`<Sidebar /> should render properly opened 1`] = `
.emotion-8 {
  list-style-type: none;
  padding: 0;
  margin: 0;
}

.emotion-1 {
  border-bottom: 1px solid transparent;
}

.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 12px;
  font-size: 15px;
  line-height: 18px;
  background-color: #3b3b3b;
  color: #73b61c;
}

.emotion-0 svg {
  margin-right: 12px;
}

.emotion-0:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #73b61c;
  background-color: #3b3b3b;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 12px;
  font-size: 15px;
  line-height: 18px;
  background-color: none;
  color: #808080;
}

.emotion-2 svg {
  margin-right: 12px;
}

.emotion-2:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #e0e0e0;
  background-color: #3b3b3b;
}

.emotion-9 {
  width: 240px;
  position: fixed;
  height: calc(100% - 44px);
  overflow-y: auto;
  z-index: 90;
  background-color: #4a4a4a;
  padding-bottom: 32px;
  -webkit-transition: left ease-in-out .3s,visibility ease-in-out .3s;
  transition: left ease-in-out .3s,visibility ease-in-out .3s;
  overscroll-behavior: contain;
  left: 0;
  top: 44px;
  padding-top: 12px;
}

.emotion-10 {
  position: fixed;
  top: 0;
  left: 0;
  bottom: 0;
  z-index: 80;
  background-color: #000000;
  opacity: 0.5;
  width: 100%;
  -webkit-transition: opacity ease-in-out .3s,width ease-in 0s 0s;
  transition: opacity ease-in-out .3s,width ease-in 0s 0s;
}

<div>
  <nav
    className="emotion-9"
  >
    <ul
      className="emotion-8"
    >
      <li
        className="emotion-1"
      >
        <a
          className="emotion-0"
          href="/76917058-b147-4556-987a-0a0f11e46d9b/jobs"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <svg
            aria-hidden={true}
            data-file-name="queue.svg"
            width={20}
          />
          Job Queue
        </a>
      </li>
      <li
        className="emotion-1"
      >
        <a
          className="emotion-2"
          href="/76917058-b147-4556-987a-0a0f11e46d9b/api-keys"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <svg
            aria-hidden={true}
            data-file-name="key.svg"
            width={20}
          />
          API Keys
        </a>
      </li>
      <li
        className="emotion-1"
      >
        <a
          className="emotion-2"
          href="/76917058-b147-4556-987a-0a0f11e46d9b/users"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <svg
            aria-hidden={true}
            data-file-name="users.svg"
            width={20}
          />
          User Admin
        </a>
      </li>
      <li
        className="emotion-1"
      >
        <a
          className="emotion-2"
          href="/76917058-b147-4556-987a-0a0f11e46d9b/account"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <svg
            aria-hidden={true}
            data-file-name="gear.svg"
            width={20}
          />
          User Admin
        </a>
      </li>
    </ul>
  </nav>
  <div
    aria-label="Close Sidebar Menu"
    className="emotion-10"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="-1"
  />
</div>
`;
